streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -67,58 +69,148 @@
 
     private static final String TOSTREAM_NAME = "KTABLE-TOSTREAM-";
 
+    public static final String STATE_STORE_NAME = "STATE-STORE-";
+
     private final ProcessorSupplier<?, ?> processorSupplier;
 
-    private final String storeName;
+    private final String queryableStoreName;
 
     private boolean sendOldValues = false;
+    private final Serde<K> keySerde;
+    private final Serde<V> valSerde;
 
+    public KTableImpl(KStreamBuilder topology,
+                      String name,
+                      ProcessorSupplier<?, ?> processorSupplier,
+                      Set<String> sourceNodes,
+                      final String queryableStoreName) {
+        super(topology, name, sourceNodes);
+        this.processorSupplier = processorSupplier;
+        this.queryableStoreName = queryableStoreName;
+        this.keySerde = null;
+        this.valSerde = null;
+    }
 
     public KTableImpl(KStreamBuilder topology,
                       String name,
                       ProcessorSupplier<?, ?> processorSupplier,
+                      final Serde<K> keySerde,
+                      final Serde<V> valSerde,
                       Set<String> sourceNodes,
-                      final String storeName) {
+                      final String queryableStoreName) {
         super(topology, name, sourceNodes);
         this.processorSupplier = processorSupplier;
-        this.storeName = storeName;
+        this.queryableStoreName = queryableStoreName;
+        this.keySerde = keySerde;
+        this.valSerde = valSerde;
     }
 
     @Override
-    public String getStoreName() {
-        return this.storeName;
+    public String queryableStoreName() {
+        return this.queryableStoreName;
     }
 
+
     @Override
-    public KTable<K, V> filter(Predicate<? super K, ? super V> predicate) {
+    public KTable<K, V> filter(final Predicate<? super K, ? super V> predicate) {
         Objects.requireNonNull(predicate, "predicate can't be null");
         String name = topology.newName(FILTER_NAME);
-        KTableProcessorSupplier<K, V, V> processorSupplier = new KTableFilter<>(this, predicate, false);
+        KTableProcessorSupplier<K, V, V> processorSupplier = new KTableFilter<>(this, predicate, false, null);
         topology.addProcessor(name, processorSupplier, this.name);
 
-        return new KTableImpl<>(topology, name, processorSupplier, sourceNodes, this.storeName);
+        return new KTableImpl<>(topology, name, processorSupplier, sourceNodes, this.queryableStoreName);
+    }
+
+    private KTable<K, V> doQueryableFilter(final Predicate<? super K, ? super V> predicate,
```

Comment:
Could we collapse the code path for having a queryable store name or not into the same function? For example:

```
filter(.. /*nothing*/) calls filter(.. (String) null);

filter(.. "storeName") calls filter(.. storeSupplier); // if storeName is not null, otherwise pass null as well

filter(.. supplier) do the actual impl, which checks if supplier is null or not
```